core/search/terms_filter.cpp

```diff
@@ -102,54 +97,72 @@ void collect_terms(
   }
 }
 
-}
+}  // namespace
 
 namespace iresearch {
 
-/*static*/ void by_terms::visit(
-    const sub_reader& segment,
-    const term_reader& field,
-    const by_terms_options::search_terms& terms,
-    filter_visitor& visitor) {
+/*static*/ void by_terms::visit(const sub_reader& segment,
+                                const term_reader& field,
+                                const by_terms_options::search_terms& terms,
+                                filter_visitor& visitor) {
   ::visit(segment, field, terms, visitor);
 }
 
-filter::prepared::ptr by_terms::prepare(
-    const index_reader& index,
-    const Order& order,
-    score_t boost,
-    const attribute_provider* /*ctx*/) const {
-  boost *= this->boost();
-  const auto& terms = options().terms;
+filter::prepared::ptr by_terms::prepare(const index_reader& index,
+                                        const Order& order, score_t boost,
+                                        const attribute_provider* ctx) const {
+  const auto& [terms, min_match, merge_type] = options();
   const size_t size = terms.size();
+  boost *= this->boost();
 
-  if (0 == size) {
+  if (0 == size || min_match > size) {
+    // Empty or unreachable search criteria
```

Comment:
What if size == 0, but min_match == 0 too?

Reply:
The query will return all documents in that case which is expectedI believe.